File: monadic-composition-async.dart
import 'dart:async';
import 'types.dart';

MathFunctionOneParamFutureOr add5 = (x) => x + 5;
MathFunctionOneParamFutureOr double = (x) => x * 2;
MathFunctionOneParamFutureOr deduct3 = (x) => x - 3;

MathFunctionOneParamFuture asycOp = (x) async {
  Completer<int> c = new Completer();
  c.complete(x + 100);
  return c.future;
};

final typeToFutureType = <T>(T x) async {
  Completer<T> c = new Completer();
  c.complete(x);
  return c.future;
};

final MathFunctionOneParamComposeListMapFuture compose = (fns) => (x) async {
      return fns.fold(typeToFutureType<int>(x), (acc, fn) async {
        final int iX = await acc;
        return fn(iX);
      });
    };

void main(List<String> args) {
  final app = compose([add5, double, deduct3, asycOp]);
  (app(5) as Future<int>).then((value) => print(value));
}

File: monadic-composition-async.ts
import { MathFuncOneParam , MathFuncCompose, MathFuncOneParamPromise, Compose} from "./types"

const add5:MathFuncOneParam = (x) => x + 5
const double:MathFuncOneParam = (x) =>x * 2
const deduct3:MathFuncOneParam =(x) => x - 3
const asyncOp:MathFuncOneParamPromise = x => Promise.resolve( x + 100)

const compose:Compose = (...fns) => x =>{
	return fns.reduce(async(accP, fn)=>{
		return accP.then((acc:number )=> fn(acc))
	}, Promise.resolve(x))
}

const app = compose(add5, double, deduct3, asyncOp)

app(4).then((res)=>console.log(res))